packages/kuma-gui/src/app/me/sources.ts

```diff
@@ -2,19 +2,49 @@ import merge from 'deepmerge'
 
 import { defineSources } from '../application/services/data-source'
 
+type Notification = {
+  timestamp: number
+  uri: string
+}
 type Storage = {
-  get: (key: string, d?: object) => Promise<object>
-  set: (key: string, value: object) => Promise<object>
+  get(key: '~notifications', d?: Notification[]): Promise<Notification[]>
+  set(key: '~notifications', value: Notification[]): Promise<Notification[]>
+  get(key: string, d?: object): Promise<object>
+  set(key: string, value: object): Promise<object>
```

Comment:
Hmm 🤔 I'd suggest to use generics for this, i.e. like

```suggestion
  get<T = object>(key: string, d?: T): Promise<T>
  set<T = object>(key: string, value: T): Promise<T>
```

and further down we could do

```ts
const dismissed = await get<Notification[]>('~notifications', [])
```

Reply:
Hmm, maybe 🤔 

What's then stopping me from doing `await get<Notification[]>('~not-a-notification-shaped-thing', [])`?

I kinda want to say, we control the `~notifications` namespace, and whenever you access that exact namespace (get or set), this is the shape you use/get. With the generic approach I lose that control I think?

Are there any other benefits to the generic approach over the overloading that you can think of here?

Reply:
>> What's then stopping me from doing await get<Notification[]>('~not-a-notification-shaped-thing', [])?

> Nothing 😄 but why would you do that?

What about this? 

```js
await get('~notifications', [])
```

(ie. forget)

> Also is there a guarantee that what is being returned is what we expect? What if it's a remote API? Maybe a type guard would be better?

> What if it's a remote API?

It's not a remote API, but then we would have auto-generated types so I wouldn't have to do anything type-wise.

> Maybe a type guard would be better?

I get the feeling we are overthinking this a little, we don't add type guards to our local storage access anywhere else, this isn't a remote API, and in my opinion we don't have scalability issues with our localStorage properties/access.

Honestly my feeling is, either approach is fine.

But, the overload approach means I can set this in one place and I get type checking on all the places in this file where I access `~notifications` which is exactly what I want. Generics doesn't give me that because I have to specify the type in multiple places. I don't at the moment see any benefits to make it a generic instead because I always want any access to `~notifications` to use the same type. i.e.

```typescript
type LocalStorage = {
  notifications: Notification[]
  [key: string]: object
}
```

Reply:
> > > What's then stopping me from doing await get<Notification[]>('~not-a-notification-shaped-thing', [])?
> 
> > Nothing 😄 but why would you do that?
> 
> What about this?
> 
> ```js
> await get('~notifications', [])
> ```
> 
> (ie. forget)

Further use would not be typed and things you try to access would cause typescript errors 🤔 

> > Also is there a guarantee that what is being returned is what we expect? What if it's a remote API? Maybe a type guard would be better?
> 
> > What if it's a remote API?
> 
> It's not a remote API, but then we would have auto-generated types so I wouldn't have to do anything type-wise.

Good point 👍 

> > Maybe a type guard would be better?
> 
> I get the feeling we are overthinking this a little, we don't add type guards to our local storage access anywhere else, this isn't a remote API, and in my opinion we don't have scalability issues with our localStorage properties/access.
> 
> Honestly my feeling is, either approach is fine.
> 
> But, the overload approach means I can set this in one place and I get type checking on all the places in this file where I access `~notifications` which is exactly what I want. Generics doesn't give me that because I have to specify the type in multiple places. I don't at the moment see any benefits to make it a generic instead because I always want any access to `~notifications` to use the same type. i.e.
> 
> ```ts
> type LocalStorage = {
>   notifications: Notification[]
>   [key: string]: object
> }
> ```

Yeah, maybe I'm overthinking it. I guess we can leave it with the overloading for now and come back here in case it becomes cumbersome. My initial comment got resolved, so I'm fine either way now 🙂

Reply:
Ok sounds good lemme know if there is anything else here, will resolve this one